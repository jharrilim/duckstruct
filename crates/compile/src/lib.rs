use ast::Root;
use codegen::{js::JsGenerator, CodeGenerator};
use hir::lower;
use parser::parse;
use tycheck::TyCheck;

pub enum TargetLang {
  Javascript,
}

pub struct Compiler;

impl Compiler {
  pub fn new() -> Self {
    Self {}
  }

  pub fn compile_js(&self, source: &str) -> Result<String, String> {
    let parse = parse(source);
    if !parse.errors.is_empty() {
      return Err(format!("{:?}", parse.errors));
    }
    let ast = match Root::cast(parse.syntax()) {
      Some(ast) => ast,
      None => return Err("Failed to generate AST from source".to_string()),
    };
    let hir = lower(ast);
    let tycheck = TyCheck::new(hir);

    Ok(JsGenerator::new(&tycheck).generate())
  }

  pub fn compile_file(
    &self,
    path: std::path::PathBuf,
    target: TargetLang
  ) -> Result<(), String> {
    let output_path = path.with_extension("js");
    let source = match std::fs::read_to_string(path) {
      Ok(source) => source,
      Err(e) => return Err(format!("Failed to read file: {}", e)),
    };
    let code = match match target {
      TargetLang::Javascript => self.compile_js(&source),
    } {
      Ok(code) => code,
      Err(err) => return Err(err),
    };

    // write to file with .ds file extension
    match std::fs::write(output_path, code) {
      Ok(()) => Ok(()),
      Err(err) => Err(format!("Failed to write to file: {}", err)),
    }
  }
}
