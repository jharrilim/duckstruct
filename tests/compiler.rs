use std::path::{Path, PathBuf};

use compile::Compiler;
use insta;
use serde::Serialize;

pub fn codestub(path: &str) -> PathBuf {
  let root = env!("CARGO_MANIFEST_DIR");
  Path::new(root).join("tests/codestubs").join(path)
}

#[derive(Serialize)]
struct DebugInfo {
  file: &'static str,
}

macro_rules! code_snapshot {
  ($file:expr) => {
    let file_path = codestub($file);
    let debug_info = DebugInfo {
      file: $file,
    };
    let code = std::fs::read_to_string(file_path).unwrap();
    let output = Compiler::new().compile_js(&code);
    assert!(output.is_ok());
    let output = output.unwrap();
    insta::with_settings!({
      info => &debug_info,
    }, {
      insta::assert_snapshot!(&output);
    });
  };
}

#[test]
pub fn compile_simple1() {
  code_snapshot!("simple1.ds");
}

#[test]
pub fn compile_factorial() {
  // code_snapshot!("factorial.ds");
}

#[test]
pub fn compile_lambda_calculus() {
  code_snapshot!("lambda_calculus.ds");
}
