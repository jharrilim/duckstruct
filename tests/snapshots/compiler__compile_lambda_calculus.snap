---
source: tests/compiler.rs
expression: "&output"
info:
  file: lambda_calculus.ds
---
const I = x => x;
const TRUE = x => y => x;
const FALSE = x => y => y;
const IF = cond => x => y => cond;
const hi = 'Hello world';
